picard/config.py

```diff
@@ -341,12 +319,51 @@ class ListOption(Option):
 setting = None
 persist = None
 
+PURGE_INTERVAL = 60000
+_thread_configs = {}
+_thread_config_lock = threading.RLock()
+
 
-def _setup(app, filename=None):
+def setup_config(app, filename=None):
     global config, setting, persist
     if filename is None:
         config = Config.from_app(app)
     else:
         config = Config.from_file(app, filename)
+    _thread_configs[threading.get_ident()] = config
     setting = config.setting
     persist = config.persist
+    QtCore.QTimer.singleShot(PURGE_INTERVAL, purge_config_instances)
+
+
+def get_config():
+    """Returns a config object for the current thread.
+
+    Config objects for threads are
+    """
+    thread_id = threading.get_ident()
+    thread_config = _thread_configs.get(thread_id)
+    if not thread_config:
+        _thread_config_lock.acquire()
+        try:
+            config_file = config.fileName()
+            log.debug('Instantiating Config for thread %s using %s.', thread_id, config_file)
+            thread_config = Config.from_file(None, config_file)
+            _thread_configs[thread_id] = thread_config
+        finally:
+            _thread_config_lock.release()
+    return thread_config
+
+
+def purge_config_instances():
+    """Removes cached config instances for no longer active threads."""
+    _thread_config_lock.acquire()
+    try:
+        all_threads = set([thread.ident for thread in threading.enumerate()])
+        for thread_id in list(_thread_configs.keys()):
+            if thread_id not in all_threads:
+                log.debug('Purging config instance for thread %s.', thread_id)
+                del _thread_configs[thread_id]
```

Comment:
This can perhaps be rewriiten so it uses only sets:

```python
        all_threads = set([thread.ident for thread in threading.enumerate()])
        threads_config = set(_thread_configs)
        for thread_id in threads_config.difference(all_threads):
            log.debug('Purging config instance for thread %s.', thread_id)
            del _thread_configs[thread_id]
```

Reply:
Not sure, I don't really think that it is easier to read.

Reply:
Yup, not easier to read for sure, but likely faster. You can still rewrite `list(_thread_configs.keys())` to `list(_thread_configs))`.